### PR TITLE
graphql/enum types added and tests modified

### PR DESCRIPTION
**Pull request by shahryart**
files in graphql/ changed to support enum types. 
example/minimal/main.go changed to test for enums.
tests in end_to_end_test.go are modified however they do not check for mapped values yet.

### CODE REVIEW

File: client/.babelrc [hunk 1]
```diff
@@ -1,11 +1,13 @@
 {
```
Comment (jellevandenhooff) [inline client/.babelrc]: are these changes necessary? if not, can we drop them?

File: client/.babelrc [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
   "presets": [
```
Comment (jellevandenhooff) [inline client/.babelrc]: could you change the commit line to be "graphql: add support for enum types"? that's more like the go style guide.

File: client/npm-debug.log [hunk 3]
```diff
@@ -0,0 +1,49 @@
+0 info it worked if it ends with ok
```
Comment (jellevandenhooff) [inline client/npm-debug.log]: let's not commit this debug log

File: example/minimal/main.go [hunk 4]
```diff
@@ -13,9 +13,11 @@ import (
 type Server struct {
 }
 
+type messageEnumType int32
```
Comment (jellevandenhooff) [inline example/minimal/main.go]: maybe MessageType? enum in the name is a little clunky.

File: example/minimal/main.go [hunk 5]
```diff
@@ -29,15 +31,26 @@ func (s *Server) registerUser(schema *schemabuilder.Schema) {
 func (s *Server) registerQuery(schema *schemabuilder.Schema) {
 	object := schema.Query()
 
-	object.FieldFunc("users", func(ctx context.Context) ([]*User, error) {
+	var tmp messageEnumType
+	schema.EnumReg(tmp, map[string]interface{}{
```
Comment (jellevandenhooff) [inline example/minimal/main.go]: maybe schema.Enum instead of schema.EnumReg, so it matches schema.Object and schema.Query?

File: example/minimal/main.go [hunk 6]
```diff
@@ -46,8 +59,11 @@ func (s *Server) registerQuery(schema *schemabuilder.Schema) {
 func (s *Server) registerMutation(schema *schemabuilder.Schema) {
 	object := schema.Mutation()
 
-	object.FieldFunc("echo", func(ctx context.Context, args struct{ Text string }) (string, error) {
-		return args.Text, nil
+	object.FieldFunc("echo", func(ctx context.Context, args struct {
+		Text      string
```
Comment (jellevandenhooff) [inline example/minimal/main.go]: maybe remove Text as argument

File: graphql/end_to_end_test.go [hunk 7]
```diff
@@ -29,72 +28,101 @@ type Slow struct {
 func TestPathError(t *testing.T) {
 	schema := schemabuilder.NewSchema()
 
-	type Inner struct{}
+	type enumType int32
```
Comment (jellevandenhooff) [inline graphql/end_to_end_test.go]: would you put your enum tests in a new TestEnum function? TestPathError is explicitly for testing PathErrors.

File: graphql/end_to_end_test.go [hunk 8]
```diff
@@ -29,72 +28,101 @@ type Slow struct {
...
+		"inner": enumType(1),
+	}, val)
+
+	e2 := graphql.Executor{}
+	val, err = e2.Execute(context.Background(), builtSchema.Query, nil, q3)
+	assert.Nil(t, err)
+	assert.Equal(t, map[string]interface{}{
+		"inner2": enumType2(1.2),
+	}, val)
+	// if err != nil || val != int32(1) {
```
Comment (jellevandenhooff) [inline graphql/end_to_end_test.go]: please don't leave commented out code here.

File: graphql/executor.go [hunk 9]
```diff
@@ -283,6 +287,14 @@ func (e *Executor) execute(ctx context.Context, typ Type, source interface{}, se
 	switch typ := typ.(type) {
 	case *Scalar:
 		return unwrap(source), nil
+	case *Enum:
+		val := unwrap(source)
+		for key, mapVal := range typ.Mapping {
```
Comment (jellevandenhooff) [inline graphql/executor.go]: could you precompute this reverse map? having to enumerate here could be expensive for enums with lots of values.

File: graphql/executor.go [hunk 10]
```diff
@@ -283,6 +287,14 @@ func (e *Executor) execute(ctx context.Context, typ Type, source interface{}, se
...
+			if val == mapVal {
+				return key, nil
+			}
+		}
+		return unwrap(source), nil
```
Comment (jellevandenhooff) [inline graphql/executor.go]: if there is no known value, should we error instead?

File: graphql/executor.go [hunk 11]
```diff
@@ -283,6 +287,14 @@ func (e *Executor) execute(ctx context.Context, typ Type, source interface{}, se
 	switch typ := typ.(type) {
 	case *Scalar:
 		return unwrap(source), nil
+	case *Enum:
+		val := unwrap(source)
```
Comment (jellevandenhooff) [inline graphql/executor.go]: is this unwrap here necessary? I think is. Let's make sure you have a test where a query returns a *Enum.

File: graphql/introspection/introspection.go [hunk 12]
```diff
@@ -190,7 +193,19 @@ func (s *introspection) registerType(schema *schemabuilder.Schema) {
...
+	object.FieldFunc("enumValues", func(t Type, args struct {
+		IncludeDeprecated *bool
+	}) []EnumValue {
+
+		switch t := t.Inner.(type) {
+		case *graphql.Enum:
+			var enumVals []EnumValue
+			for _, v := range t.Values {
+				enumVals = append(enumVals,
+					EnumValue{Name: v, Description: "", IsDeprecated: false, DeprecationReason: ""})
```
Comment (jellevandenhooff) [inline graphql/introspection/introspection.go]: How about we include the Go value in the description?

File: graphql/parser.go [hunk 13]
```diff
@@ -33,6 +33,9 @@ func valueToJson(value ast.Value, vars map[string]interface{}) (interface{}, err
 		return value.Value, nil
 	case *ast.BooleanValue:
 		return value.Value, nil
+	case *ast.EnumValue:
+		//BOX IT UP HERE (SAME AS THE BOX THAT GETS THE TYPE IN FIELD FUNC)
```
Comment (jellevandenhooff) [inline graphql/parser.go]: i don't understand this comment. could you rewrite it or delete it?

File: graphql/parser.go [hunk 14]
```diff
@@ -379,7 +382,6 @@ func Parse(source string, vars map[string]interface{}) (*Query, error) {
 		Kind:         kind,
 		SelectionSet: nil,
 	}
-
 	// Parse variable definitions, default values, etc.
```
Comment (jellevandenhooff) [inline graphql/parser.go]: why get rid of these newlines? makes your diff larger than necessary

File: graphql/schemabuilder/reflect.go [hunk 15]
```diff
@@ -170,6 +170,27 @@ func getScalarArgParser(typ reflect.Type) (*argParser, graphql.Type, bool) {
...
+			return errors.New("not a string")
+		}
+		val, ok := sb.enumMappings[typ][asString]
+		if !ok {
+			return errors.New("not an enum")
+		}
+		dest.Set(reflect.ValueOf(val).Convert(dest.Type()))
+		return nil
+	}}, &graphql.Enum{Type: typ.Name(), Values: values, Mapping: sb.enumMappings[typ]}, nil
+	//&graphql.Scalar{Type: reflect.TypeOf(sb.enumMappings[typ]).Elem().Name()}, nil
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: don't leave comments

File: graphql/schemabuilder/reflect.go [hunk 16]
```diff
@@ -170,6 +170,27 @@ func getScalarArgParser(typ reflect.Type) (*argParser, graphql.Type, bool) {
...
+		values = append(values, mapping)
+	}
+	return &argParser{FromJSON: func(value interface{}, dest reflect.Value) error {
+		asString, ok := value.(string)
+		if !ok {
+			return errors.New("not a string")
+		}
+		val, ok := sb.enumMappings[typ][asString]
+		if !ok {
+			return errors.New("not an enum")
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: maybe "unknown enum value '%s'"?

File: graphql/schemabuilder/reflect.go [hunk 17]
```diff
@@ -182,30 +203,37 @@ type argField struct {
 	optional bool
 }
 
-func makeArgParser(typ reflect.Type) (*argParser, graphql.Type, error) {
+//enummappings could be wrapped into schema info or something
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: i think this comment is addressed as well

File: graphql/schemabuilder/reflect.go [hunk 18]
```diff
@@ -182,30 +203,37 @@ type argField struct {
...
+	parser, argType, err := sb.makeArgParserInner(typ)
 	if err != nil {
 		return nil, nil, err
 	}
 	return parser, &graphql.NonNull{Type: argType}, nil
 }
 
-func makeArgParserInner(typ reflect.Type) (*argParser, graphql.Type, error) {
+func (sb *schemaBuilder) makeArgParserInner(typ reflect.Type) (*argParser, graphql.Type, error) {
+
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: no newline here

File: graphql/schemabuilder/reflect.go [hunk 19]
```diff
@@ -239,9 +267,9 @@ func wrapPtrParser(inner *argParser) *argParser {
 	}
 }
 
-func makeStructParser(typ reflect.Type) (*argParser, graphql.Type, error) {
-	fields := make(map[string]argField)
+func (sb *schemaBuilder) makeStructParser(typ reflect.Type) (*argParser, graphql.Type, error) {
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: could you undo this whitespace change here

File: graphql/schemabuilder/reflect.go [hunk 20]
```diff
@@ -328,8 +355,8 @@ func makeStructParser(typ reflect.Type) (*argParser, graphql.Type, error) {
 	}, argType, nil
 }
 
-func makeSliceParser(typ reflect.Type) (*argParser, graphql.Type, error) {
-	inner, argType, err := makeArgParser(typ.Elem())
+func (sb *schemaBuilder) makeSliceParser(typ reflect.Type) (*argParser, graphql.Type, error) {
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: in the future, I would make the change that make make***Parser all methods its own commit instead

File: graphql/schemabuilder/reflect.go [hunk 21]
```diff
@@ -740,7 +783,8 @@ func (sb *schemaBuilder) getType(t reflect.Type) (graphql.Type, error) {
 }
 
 type Schema struct {
-	objects map[string]*Object
+	objects   map[string]*Object
+	EnumTypes map[reflect.Type]map[string]interface{}
```
Comment (jellevandenhooff) [inline graphql/schemabuilder/reflect.go]: does this need to be public?

File: graphql/end_to_end_test.go [hunk 22]
```diff
@@ -93,14 +93,14 @@ func TestPathError(t *testing.T) {
 	val, err := e.Execute(context.Background(), builtSchema.Query, nil, q2)
 	assert.Nil(t, err)
 	assert.Equal(t, map[string]interface{}{
-		"inner": enumType(1),
+		"inner": "firstField",
```
Comment (jellevandenhooff) [inline graphql/end_to_end_test.go]: when you have a later commit that changes an earlier commit, please squash them together

File: graphql/parser.go [hunk 23]
```diff
@@ -67,7 +66,7 @@ func valueToJson(value ast.Value, vars map[string]interface{}) (interface{}, err
 		}
 		return list, nil
 	default:
-		return nil, NewClientError("unsupported value type: %s", value.GetKind())
+		return nil, NewClientError("unsupported value type ????: %s", value.GetKind())
```
Comment (jellevandenhooff) [inline graphql/parser.go]: please don't leave debug comments like this (also below)

File: graphql/schemabuilder/reflect.go [hunk 24]
```diff
@@ -747,6 +794,18 @@ func NewSchema() *Schema {
 	}
 }
 
+func (s *Schema) Enum(val interface{}, enumMap map[string]interface{}) { //could be a type or a value
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: nit: clarify this comment/make it a sentence?

File: graphql/schemabuilder/reflect.go [hunk 25]
```diff
@@ -170,6 +170,26 @@ func getScalarArgParser(typ reflect.Type) (*argParser, graphql.Type, bool) {
...
+		values = append(values, mapping)
+	}
+	return &argParser{FromJSON: func(value interface{}, dest reflect.Value) error {
+		asString, ok := value.(string)
+		if !ok {
+			return errors.New("not a string")
+		}
+		val, ok := sb.enumMappings[typ].Map[asString]
+		if !ok {
+			return errors.New("unknown enum value " + asString)
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: `fmt.Errorf`

File: graphql/types.go [hunk 26]
```diff
@@ -26,6 +26,23 @@ func (s *Scalar) String() string {
 	return s.Type
 }
 
+//Enum is a leaf value
```
Comment (stephen) [inline graphql/types.go]: nit: comment sentence / space at beginning

File: graphql/introspection/introspection.go (hunk 12 above)
Comment (stephen) [reply graphql/introspection/introspection.go]: this seems like a nice thing to do

File: graphql/end_to_end_test.go [hunk 27]
```diff
@@ -98,6 +98,129 @@ func TestPathError(t *testing.T) {
...
+	schema.Enum(enumType2(1.2), map[string]interface{}{
+		"this": enumType2(1.2),
+		"is":   enumType2(3.2),
+		"a":    enumType2(4.3),
+		"map":  enumType2(5.3),
+	})
+
+	query := schema.Query()
+	query.FieldFunc("inner", func(args struct {
+		EnumField enumType //will be parsed as the type that gets passed in here
```
Comment (stephen) [inline graphql/end_to_end_test.go]: nit: comment sentence / put it on its own line

File: example/minimal/main.go [hunk 28]
```diff
@@ -29,15 +31,26 @@ func (s *Server) registerUser(schema *schemabuilder.Schema) {
...
-	object.FieldFunc("users", func(ctx context.Context) ([]*User, error) {
+	var tmp MessageType
+	schema.Enum(tmp, map[string]interface{}{
+		"first":  MessageType(1),
+		"second": MessageType(2),
+		"third":  MessageType(3),
+	})
+
+	object.FieldFunc("users", func(ctx context.Context, args struct {
+		EnumField MessageType
```
Comment (stephen) [inline example/minimal/main.go]: since this is the example code, let's make this more semantically interesting? maybe a `ROle` enum for `user` / `manager` / `administrator` or something

File: graphql/schemabuilder/reflect.go [hunk 29]
```diff
@@ -749,6 +767,21 @@ func NewSchema() *Schema {
 	}
 }
 
+<<<<<<< HEAD
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: eek

i think this change doesn't belong here

File: graphql/schemabuilder/reflect.go [hunk 30]
```diff
@@ -170,6 +170,26 @@ func getScalarArgParser(typ reflect.Type) (*argParser, graphql.Type, bool) {
...
+		}
+		val, ok := sb.enumMappings[typ].Map[asString]
+		if !ok {
+			return fmt.Errorf("unknown enum value %v", asString)
+		}
+		dest.Set(reflect.ValueOf(val).Convert(dest.Type()))
+		return nil
+	}, Type: typ}, &graphql.Enum{Type: typ.Name(), Values: values, ReverseMap: sb.enumMappings[typ].ReverseMap}, nil
+
+}
```
Comment (stephen) [inline graphql/schemabuilder/reflect.go]: i think this doesn't belong in this commit?

File: graphql/schemabuilder/reflect_test.go [hunk 31]
```diff
@@ -459,15 +468,15 @@ func TestArgParser(t *testing.T) {
...
 	}
 
-	if _, _, err := makeArgParser(reflect.TypeOf(&anonymous{})); err == nil {
+	if _, _, err := sb.makeArgParser(reflect.TypeOf(&anonymous{})); err == nil {
 		t.Error("expected anonymous fields to fail")
 	}
 
-	if _, _, err := makeArgParser(reflect.TypeOf(&unsupported{})); err == nil {
+	if _, _, err := sb.makeArgParser(reflect.TypeOf(&unsupported{})); err == nil {
 		t.Error("expected unsupported fields to fail")
```
Comment (stephen) [inline graphql/schemabuilder/reflect_test.go]: this should be in the previous commit.
